verl/utils/reward_score/coder1/__init__.py

```diff
@@ -0,0 +1,155 @@
+# SPDX-FileCopyrightText: (c) iSE UIUC Research Group
+#
+# SPDX-License-Identifier: Apache-2.0
+
+import re
+from typing import Tuple, Optional
+import time
+import json
+from concurrent.futures import ThreadPoolExecutor, as_completed
+
+import numpy as np
+
+from verl.utils.reward_score.coder1.firejail import code_exec_firejail as code_exec, _ERROR_MSG_PREFIX
+
+_MAX_CHAR_DISPLAY = 2048
+
+
+def exec_check_stdio(code, stdin, stdout, timeout=None):
+    succ, output = code_exec(code=code, stdin=stdin, timeout=timeout)
+    return succ, output, stdin, stdout
+
+
+def validate_response_structure(processed_str: str) -> bool:
+    think_pattern = re.compile(r"<think>(.*?)</think>", re.DOTALL)
+    if match := think_pattern.match(processed_str):
+        thinking = f"<think>{match.group(1)}</think>"
+        response = (response.split(thinking)[-1].strip().strip("<answer>").strip("</answer>").strip())
+        return ("<answer>" not in response and "</answer>" not in response and "<think>" not in response and
+                "</think>" not in response)
+
+    return False
+
+
+def try_extract_answer(response: str) -> Tuple[Optional[str], str]:
+    answer_pattern = r"<think>(.*?)</think>"
+    if matches := list(re.finditer(answer_pattern, response, re.DOTALL)):
+        thinking = f"<think>{matches[-1].group(1)}</think>"
+        response = response.split(thinking)[-1].strip()
+    return response.strip("<answer>").strip("</answer>").strip()
+
+
+def extract_code_from_string(solution_str):
+    pattern = re.compile(r'```(?:\w+)?\n(.*?)\n```', re.DOTALL)
+    code_blocks = pattern.findall(solution_str)
+    return '\n'.join(code_blocks).strip()
+
+
+def _check_fmt(response: str) -> Tuple[bool, str]:
+    reward_log = ("-" * 16 + "Bad format detected!" + "-" * 16 + "\n" + "-" * 16 + "Original Model Output" + "-" * 16 +
+                  "\n" + response)
+
+    if not validate_response_structure(response):
+        return False, reward_log
+
+    if len(extract_code_from_string(try_extract_answer(response)).strip()) == 0:
+        return False, reward_log
+
+    return True, ""
+
+
+def _check_correctness(solution_str, ground_truth, extra_info) -> Tuple[bool, str]:
```

Comment:
could you add some comments on the high level approach of how correctness is checked?